### PR TITLE
Meterpreter interactive shell

### PR DESCRIPTION
**Pull request by dwelch-r7**
This PR adds support for a fully interactive meterpreter experience (currently only tested with the python meterpreter on linux)
What I mean by fully interactive is that all the creature comforts of a normal shell are available to you, i.e. Left/Right arrow keys for editing your command, Up/Down arrow keys for cycling through previous commands in your history, `Ctrl + R` for reverse search and so on.

There is a companion PR coming soon to the python meterpreter aswell to allow dynamic resizing of the terminal

Definitely still a little rough around the edges but putting it up for some feedback on the general concept of how I'm going about this

Link to the companion payloads PR: https://github.com/rapid7/metasploit-payloads/pull/499

### CODE REVIEW

File: lib/rex/post/meterpreter/core_ids.rb [hunk 1]
```diff
@@ -47,6 +47,8 @@ module Meterpreter
 COMMAND_ID_CORE_TRANSPORT_SLEEP          = EXTENSION_ID_CORE + 33
 COMMAND_ID_CORE_PIVOT_SESSION_NEW        = EXTENSION_ID_CORE + 34
 
+COMMAND_ID_CORE_SET_TERM_SIZE            = EXTENSION_ID_CORE + 337
```
Comment (dwelch-r7) [inline lib/rex/post/meterpreter/core_ids.rb:50]: This is just a placeholder for now, I'm not even sure this command should exist in core or if it should be in stdapi?

File: lib/rex/post/meterpreter/ui/console.rb [hunk 2]
```diff
@@ -75,11 +75,12 @@ def interact(&block)
   #
   # Interacts with the supplied channel.
   #
-  def interact_with_channel(channel)
+  def interact_with_channel(channel, raw: false)
     channel.extend(InteractiveChannel) unless (channel.kind_of?(InteractiveChannel) == true)
     channel.on_command_proc = self.on_command_proc if self.on_command_proc
     channel.on_print_proc   = self.on_print_proc if self.on_print_proc
     channel.on_log_proc = method(:log_output) if self.respond_to?(:log_output, true)
+    channel.raw = raw
```
Comment (dwelch-r7) [inline lib/rex/post/meterpreter/ui/console.rb:83]: I couldn;t come up with a better name than `raw` we already use interact/interactive/interacting all over the place taking suggestions :sweat_smile:

**Comment by adfoster-r7** [comment]
cross-referencing https://github.com/rapid7/metasploit-framework/issues/14369

**Comment by adfoster-r7** [comment]
We'll probably want to have an answer to the security considerations of handling control codes too: https://marc.info/?l=bugtraq&m=104612710031920&w=2

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 3]
```diff
@@ -302,6 +319,8 @@ def cmd_shell(*args)
         end
 
         return true
+      when '-i'
+        raw = true
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:323]: TODO: We'll want to add validation for ensuring that raw mode needs to be coupled with a pty

File: lib/msf/core/feature_manager.rb [hunk 4]
```diff
@@ -66,12 +72,19 @@ def set(name, value)
...
           Rex::Text::Table.wrap_tables!
         else
           Rex::Text::Table.unwrap_tables!
         end
+      when FULLY_INTERACTIVE_SHELLS
+        if value
+          Rex::Post::Meterpreter::Ui::Console::CommandDispatcher::Stdapi::Sys.enable_fully_interactive!
+        else
+          Rex::Post::Meterpreter::Ui::Console::CommandDispatcher::Stdapi::Sys.disable_fully_interactive!
+        end
```
Comment (adfoster-r7) [inline lib/msf/core/feature_manager.rb:87]: Let's look at a way of doing this without the side effects, so we don't need to load all of the meterpreter classes when loading the feature manager :+1:

File: lib/rex/post/meterpreter/ui/console/interactive_channel.rb [hunk 5]
```diff
@@ -14,15 +14,24 @@ module Console::InteractiveChannel
...
   #
   def _interact
     # If the channel has a left-side socket, then we can interact with it.
     if (self.lsock)
       self.interactive(true)
-
-      interact_stream(self)
+      if raw
+        _winch
+        handle_winch
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/interactive_channel.rb:27]: We could add this to the base class for consistency

File: lib/rex/post/meterpreter/ui/console/interactive_channel.rb [hunk 6]
```diff
@@ -14,15 +14,24 @@ module Console::InteractiveChannel
...
   # Interacts with self.
   #
   def _interact
     # If the channel has a left-side socket, then we can interact with it.
     if (self.lsock)
       self.interactive(true)
-
-      interact_stream(self)
+      if raw
+        _winch
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/interactive_channel.rb:26]:
```suggestion
        set_tty_size
```

File: lib/rex/post/meterpreter/ui/console/interactive_channel.rb [hunk 7]
```diff
@@ -92,6 +105,18 @@ def _remote_fd(stream)
...
+  attr_accessor :cols
+
+  def _winch
+    rows, cols = ::IO.console.winsize
+    unless rows == self.rows && cols == self.cols
+      set_term_size(rows, cols)
+      self.rows = rows
+      self.cols = cols
+    end
+  end
```
[the remark this answers is not in the document]
Comment (adfoster-r7) [reply lib/rex/post/meterpreter/ui/console/interactive_channel.rb:118]: Not strongly opinionated on the method names

File: lib/rex/post/hwbridge/ui/console.rb [hunk 8]
```diff
@@ -75,7 +75,7 @@ def interact(&block)
   #
   # Interacts with the supplied channel.
   #
-  def interact_with_channel(channel)
+  def interact_with_channel(channel, *_)
```
Comment (adfoster-r7) [inline lib/rex/post/hwbridge/ui/console.rb:78]:
```suggestion
  def interact_with_channel(channel, **_)
```

https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 9]
```diff
@@ -19,31 +19,6 @@ class Console::CommandDispatcher::Stdapi::Sys
...
-    "-s" => [ true,  "Execute process in a given session as the session user"  ])
-
-  #
-  # Options used by the 'shell' command.
-  #
-  @@shell_opts = Rex::Parser::Arguments.new(
-    "-h" => [ false, "Help menu."                                          ],
-    "-l" => [ false, "List available shells (/etc/shells)."                ],
-    "-t" => [ true,  "Spawn a PTY shell (/bin/bash if no argument given)." ]) # ssh(1) -t
-
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:46]: Just to make the git blame cleaner, we could revert this change, and just add an extra class variable with the extra options merged in:
```
      @@shell_opts_with_fully_interactive_shell = Rex::Parser::Arguments.new(@@shell_opts.fmt.merge(
        { '-i' => [ false, 'Drop into a fully interactive shell. (Only used in conjunction with `-t`'] }
      ))
```

And update the accessor methods to reference this instead. When the feature is added for real, the git diff will just be deleting the accessor methods, and adding the `-i` to the existing `@@shell_opts` which is easier to review.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 10]
```diff
@@ -104,6 +79,49 @@ class Console::CommandDispatcher::Stdapi::Sys
...
+      '-t' => [true, 'Spawn a PTY shell (/bin/bash if no argument given).']
+    }
+    if Msf::FeatureManager.instance.enabled?(Msf::FeatureManager::FULLY_INTERACTIVE_SHELLS)
+      Rex::Parser::Arguments.new(default_shell_opts.merge(
+        { '-i' => [ false, 'Drop into a fully interactive shell. (Only used in conjunction with `-t`'] }
+      ))
+    else
+      Rex::Parser::Arguments.new(default_shell_opts)
+    end
+  end
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:98]:
```suggestion
  def shell_opts
    if framework.features.enabled?(Msf::FeatureManager::FULLY_INTERACTIVE_SHELLS)
      @@shell_opts_with_fully_interactive_shell
    else
      @shell_opts
    end
  end
```

Comment (adfoster-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:98]: The instance of the feature manager is available on `framework`

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb (hunk 3 above)
Comment (adfoster-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:333]: I'm probably missing it; Was there validation added for this? 👀

Comment (dwelch-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:333]: yup! the raw flag is now only added when the `pty_shell` function is called
https://github.com/rapid7/metasploit-framework/pull/15522/files/d98896564d25ebdfe5671d637647b14327aa2956#diff-f6355bf5fe7f8b6713ddc4f22831b4d74290abb1296dccce03289b88be417226R383

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb (hunk 10 above)
Comment (dwelch-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:98]: yea tried that before and I thought it didn;t work, so I replicated it just now
```meterpreter > shell -it
[-] Error running command shell: NameError undefined local variable or method `framework' for #<Rex::Post::Meterpreter::Ui::Console::CommandDispatcher::Stdapi::Sys:0x00007ff6e587a1c0>
```

Comment (adfoster-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:98]: Ah, for meterpreter's context it would be `client.framework.features.enabled?(...)`

Comment (dwelch-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:98]: aye that's worked, cheers

File: lib/rex/post/hwbridge/ui/console.rb (hunk 8 above)
Comment (adfoster-r7) [reply lib/rex/post/hwbridge/ui/console.rb:78]: I believe it will still work without this change, but might as well be explicit with our intent :+1:

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb [hunk 11]
```diff
@@ -44,6 +48,10 @@ class Console::CommandDispatcher::Stdapi::Sys
     "-l" => [ false, "List available shells (/etc/shells)."                ],
     "-t" => [ true,  "Spawn a PTY shell (/bin/bash if no argument given)." ]) # ssh(1) -t
 
+  @@shell_opts_with_fully_interactive_shell = Rex::Parser::Arguments.new(@@shell_opts.fmt.merge(
+    { '-i' => [ false, 'Drop into a fully interactive shell. (Only used in conjunction with `-t`'] }
```
Comment (adfoster-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/sys.rb:52]:
```suggestion
    { '-i' => [ false, 'Drop into a fully interactive shell. (Only used in conjunction with `-t`).'] }
```

**Comment by adfoster-r7** [comment]
> We'll probably want to have an answer to the security considerations of handling control codes too: https://marc.info/?l=bugtraq&m=104612710031920&w=2

Looks like this isn't a regression in behavior for meterpreter, I don't see this as a blocker for now. Especially as it's behind a feature flag.

File: lib/rex/ui/text/input/readline.rb [hunk 12]
```diff
@@ -82,6 +82,13 @@ def gets()
       end
     end
 
+    def getch()
+      begin
+        self.fd.getch
+      rescue ::Errno::EINTR
+        retry
```
Comment (adfoster-r7) [inline lib/rex/ui/text/input/readline.rb:89]: Just confirming, is this the behavior we want? 🤔

Comment (dwelch-r7) [reply lib/rex/ui/text/input/readline.rb:89]: Potentially not, but it _is_ consistent with what we already have
https://github.com/rapid7/metasploit-framework/pull/15522/files/84003678af354993c7c83f95840c885914626226#diff-c323d311fe972840e9920c1bffaaf2746018bf5faa463c2ad1672ba18e19f0dfR77-R83

**Comment by adfoster-r7** [comment]
## Release Notes

Adds support for fully interactive shells against Linux environments with `shell -it`. This functionality is behind a feature flag and can be enabled with `features set fully_interactive_shells true`
